src/charm.py

```diff
@@ -1,104 +1,400 @@
+    #  User management support functions
+    # ===================================
+
+    def _push_container_config(self, users=None) -> None:
+        """Updates config files stored on pgbouncer container and reloads application.
+
+        Updates userlist.txt and pgbouncer.ini config files, reloading pgbouncer application once
+        updated.
+
+        Params:
+            users: a dictionary of usernames and passwords
+        """
+        if users is None:
+            users = self._get_userlist_from_container()
         else:
-            event.set_results({"fortune": "A bug in the code is worth two in the documentation."})
+            self._push_userlist(users)
+
+        self._push_pgbouncer_ini(users, reload_pgbouncer=True)
+
+    def _push_pgbouncer_ini(self, users=None, reload_pgbouncer=False) -> None:
+        """Generate pgbouncer.ini from juju config and deploy it to the container.
+
+        Params:
+            users: a dictionary of usernames and passwords
+            reload_pgbouncer: A boolean defining whether or not to reload the pgbouncer application
+                in the container. When config files are updated, pgbouncer must be restarted for
+                the changes to take effect. However, these config updates can be done in batches,
+                minimising the amount of necessary restarts.
+        """
+        if users is None:
+            users = self._get_userlist_from_container()
+
+        pgb_container = self.unit.get_container(self._pgbouncer_service)
+        pgbouncer_ini = self._generate_pgbouncer_ini(users)
+
+        try:
+            # Check that we're not updating this file unnecessarily
+            if pgb_container.pull(INI_PATH).read() == pgbouncer_ini:
+                logger.info("updated config does not modify existing pgbouncer config")
+                return
+        except FileNotFoundError:
+            # There is no existing pgbouncer.ini file, so carry on and add one.
+            pass
+
+        pgb_container.push(
+            INI_PATH,
+            pgbouncer_ini,
+            user=self._pgbouncer_user,
+            permissions=0o600,
+            make_dirs=True,
+        )
+        logging.info("pushed new pgbouncer.ini config file to pgbouncer container")
+
+        if reload_pgbouncer:
+            self._reload_pgbouncer()
+
+    def _generate_pgbouncer_ini(self, users=None) -> str:
+        """Generate pgbouncer.ini from config.
+
+        This is a basic stub method, and will be updated in future to generate more complex
+        pgbouncer.ini files in a more sophisticated way.
+
+        TODO compare to real-world pgbouncer.ini files and devise a more sophisticated way of
+        generating them. Look at the previous charm's implementation of this.
+        TODO evaluate other user types, such as stat_users etc
+        TODO define/parse databases based on relation to postgres - does it need to be an exposed
+        config option?
+
+        Params:
```

Comment:
```suggestion
        Args:
```
You may want to use Args throughout your code, we follow [this styleguide](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings) for docstrings so I think you'll need to use `Args`. You'll also want to specify returns

Reply:
Type hinting has been added. Generally event handlers don't have the `event` argument added into the docstring unless that argument should contain certain parameters, such as the `username` or `password` parameters for user action management.